spec/interactions/buy_ticket_spec.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe BuyTicket do
+  let(:user) { create :user }
+  let(:event) { create :event_with_sold_tickets }
+  let(:ticket) { event.tickets[0] }
+  let(:outcome) { BuyTicket.run(ticket: ticket, user: user) }
+
+  describe 'Buy ticket' do
+    context "ticket wasn't sold and user isn't owner" do
+      it do
+        expect(outcome).to be_valid
+        expect(ticket.bought_by).to eq(user)
+      end
+    end
+
+    context 'ticket has been sold' do
+      let(:ticket) { event.tickets[3] }
```

Comment:
Fajny, dokładny test, ale mam małą uwagę.

Odnośnie tworzenia obiektów do testów w taki sposob, że `create :event_with_sold_tickets` tworzy event z biletami, z których niektóre są sprzedane, a innne nie:

Na pierwszy rzut oka nie wiadomo dlaczego `event.tickets[3]` oznacza sprzedany bilet, a `event.tickets[0]` niesprzedany.

Jak można to rozwiązać inaczej:

```ruby
RSpec.describe BuyTicket do
  let(:user) { create :user }
  let(:ticket) { create :ticket }
  let(:outcome) { BuyTicket.run(ticket: ticket, user: user) }

   describe 'Buy ticket' do
    context "ticket wasn't sold and user isn't owner" do
      it do
        expect(outcome).to be_valid
        expect(ticket.bought_by).to eq(user)
      end
    end

     context 'ticket has been sold' do
      let(:ticket) { create :ticket, :sold }
      it do
        expect(outcome).to_not be_valid
      end
       it do
        expect(outcome.errors.full_messages.join('. ')).to match(/already been sold/)
      end
    end

     context 'user buy his own ticket' do
      let(:user) { ticket.owner }
       it do
        expect(outcome).to_not be_valid
      end
       it do
        expect(outcome.errors.full_messages.join('. ')).to match(/your own ticket/)
      end
    end
  end
end
```

Reply:
Dlaczego nie jest dobrze robić factories typu `create :event_with_sold_tickets`:
przede wszystkim dlatego, że z czasem powstaną potrzeby na kolejne odmiany tego typu obiektów np:
`create :event_with_all_tickets_sold`
`create :event_with_none_ticket_sold`
itd. i rośnie nam stopień skomplikowania tego wszystkiego. Lepiej używać traits. Wtedy możemy użyć kilku modyfikatorów jednocześnie.

Do tego w tym teście nie potrzebujemy w ogóle korzystać z obiektu `event`, więc łatwiej jest stworzyć od razu `ticket`.